fido2/ctap.c

```diff
@@ -1154,6 +1157,270 @@ uint8_t ctap_get_next_assertion(CborEncoder * encoder)
     return 0;
 }
 
+uint8_t ctap_cred_metadata(CborEncoder * encoder)
+{
+    CborEncoder map;
+    int ret = cbor_encoder_create_map(encoder, &map, 2);
+    check_ret(ret);
+    ret = cbor_encode_int(&map, 1);
+    check_ret(ret);
+    ret = cbor_encode_int(&map, STATE.rk_stored);
+    check_ret(ret);
+    ret = cbor_encode_int(&map, 2);
+    check_ret(ret);
+    int remaining_rks = ctap_rk_size() - STATE.rk_stored;
+    ret = cbor_encode_int(&map, remaining_rks);
+    check_ret(ret);
+    ret = cbor_encoder_close_container(encoder, &map);
+    check_ret(ret);
+    return 0;
+}
+
+uint8_t ctap_cred_rp(CborEncoder * encoder, int rk_ind, int rp_count)
+{
+    CTAP_residentKey rk;
+    ctap_load_rk(rk_ind, &rk);
+    // SHA256 hash of "ssh:"
+    uint8_t ssh_hash[32] = {0xe3, 0x06, 0x10, 0xe8, 0xa1, 0x62, 0x11, 0x59,
+                            0x60, 0xfe, 0x1e, 0xc2, 0x23, 0xe6, 0x52, 0x9c,
+                            0x9f, 0x4b, 0x6e, 0x80, 0x20, 0x0d, 0xcb, 0x5e,
+                            0x5c, 0x32, 0x1c, 0x8a, 0xf1, 0xe2, 0xb1, 0xbf};
+
+    CborEncoder map;
+    size_t map_size = rp_count > 0 ? 3 : 2;
+    int ret = cbor_encoder_create_map(encoder, &map, map_size);
+    check_ret(ret);
+    ret = cbor_encode_int(&map, 3);
+    check_ret(ret);
+    {
+        CborEncoder rp;
+        ret = cbor_encoder_create_map(&map, &rp, 2);
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&rp, "id");
+        check_ret(ret);
+        if (memcmp(ssh_hash, rk.id.rpIdHash, 32) == 0) {
+            // recover the rpId from the hash because Solo stores only the hash :(
+            ret = cbor_encode_text_stringz(&rp, "ssh:");
+        } else {
+            ret = cbor_encode_text_stringz(&rp, "");
+        }
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&rp, "name");
+        check_ret(ret);
+        ret = cbor_encode_text_stringz(&rp, rk.user.name);
+        check_ret(ret);
+        ret = cbor_encoder_close_container(&map, &rp);
+        check_ret(ret);
+    }
+    ret = cbor_encode_int(&map, 4);
+    check_ret(ret);
+    cbor_encode_byte_string(&map, rk.id.rpIdHash, 32);
+    check_ret(ret);
+    if (rp_count > 0)
+    {
+        ret = cbor_encode_int(&map, 5);
```

Comment:
From the spec, it looks like this isn't returned on the "next" commands, i.e. only when `rk_ind == 0`

Reply:
rp_count is greater than zero only when the command is `CM_cmdRPBegin`, see line 1367